File: abilities-common/src/main/java/io/github/solusmods/eternalcore/abilities/impl/network/c2s/RequestAbilityTogglePacket.java
package io.github.solusmods.eternalcore.abilities.impl.network.c2s;

import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import io.github.solusmods.eternalcore.abilities.EternalCoreAbilities;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;

public record RequestAbilityTogglePacket(
        ResourceLocation abilityId
) implements CustomPacketPayload {
    public static final CustomPacketPayload.Type<RequestAbilityTogglePacket> TYPE = new CustomPacketPayload.Type<>(EternalCoreAbilities.create("request_ability_toggle"));
    public static final StreamCodec<FriendlyByteBuf, RequestAbilityTogglePacket> STREAM_CODEC = CustomPacketPayload.codec(RequestAbilityTogglePacket::encode, RequestAbilityTogglePacket::new);

    public RequestAbilityTogglePacket(FriendlyByteBuf buf) {
        this(buf.readResourceLocation());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeResourceLocation(this.abilityId);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.SERVER) return;
        context.queue(() -> ClientAccess.handle(this, context.getPlayer()));
    }

    public @NotNull Type<RequestAbilityTogglePacket> type() {
        return TYPE;
    }
}

File: abilities-common/src/main/kotlin/io/github/solusmods/eternalcore/abilities/impl/network/c2s/RequestAbilityTogglePacket.kt
package io.github.solusmods.eternalcore.abilities.impl.network.c2s

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.abilities.EternalCoreAbilities
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamDecoder
import net.minecraft.network.codec.StreamMemberEncoder
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation

@JvmRecord
data class RequestAbilityTogglePacket(
    val abilityId: ResourceLocation?
) : CustomPacketPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readResourceLocation())

    fun encode(buf: FriendlyByteBuf) {
        buf.writeResourceLocation(this.abilityId)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.getEnvironment() != Env.SERVER) return
        context.queue(Runnable { ClientAccess.handle(this, context.getPlayer()) })
    }

    override fun type(): CustomPacketPayload.Type<RequestAbilityTogglePacket?> {
        return TYPE
    }

    companion object {
        val TYPE: CustomPacketPayload.Type<RequestAbilityTogglePacket?> =
            CustomPacketPayload.Type<RequestAbilityTogglePacket?>(EternalCoreAbilities.create("request_ability_toggle"))
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf?, RequestAbilityTogglePacket?> =
            CustomPacketPayload.codec<FriendlyByteBuf?, RequestAbilityTogglePacket?>(
                StreamMemberEncoder { obj: RequestAbilityTogglePacket?, buf: FriendlyByteBuf? -> obj!!.encode(buf!!) },
                StreamDecoder { buf: FriendlyByteBuf? ->
                    RequestAbilityTogglePacket(
                        buf!!
                    )
                })
    }
}
